File: android_samples_java/22.1 LocationDemo/app/src/main/java/com/msnegi/locationdemo/LocManager.java
package com.msnegi.locationdemo;

import android.app.Activity;
import android.content.Context;
import android.location.Location;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationListener;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;

import java.text.DateFormat;
import java.util.Date;

public class LocManager implements LocationListener, GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {

    private static final String TAG = LocManager.class.getSimpleName();
    private static LocManager instance;
    public GoogleApiClient mGoogleApiClient;
    private Context mContext;
    //public static final int LOCATION_UPDATE_INTERVAL = 3 * 60 * 1000;    // after 10 mints
    //public static final int LOCATION_FASTEST_UPDATE_INTERVAL = 50 * 1000;
    public static final int PLAY_SERVICES_RESOLUTION_REQUEST = 9000;
    private static final long INTERVAL = 1000 * 10;
    private static final long FASTEST_INTERVAL = 1000 * 5;  //benificial in case of services only
    LocationRequest mLocationRequest;
    public Location mCurrentLocation;
    String mLastUpdateTime = "";
    LatLongReceived latlongRec;

    public static LocManager getInstance() {
        if (instance == null) {
            instance = new LocManager();
        }
        return instance;
    }

    public void connectGoogleClient(Context context, LatLongReceived latlongRec) {
        this.mContext = context;
        this.latlongRec = latlongRec;
        if (isGooglePlayServiceAvailable(context)) {
            buildGoogleApiClient(context);
        }
    }

    public static boolean isGooglePlayServiceAvailable(Context context) {
        return checkPlayServices(context);
    }

    public static boolean checkPlayServices(Context context) {
        GoogleApiAvailability gApi = GoogleApiAvailability.getInstance();
        int resultCode = gApi.isGooglePlayServicesAvailable(context);
        if (resultCode != ConnectionResult.SUCCESS) {
            if (gApi.isUserResolvableError(resultCode)) {
                gApi.getErrorDialog((Activity) context, resultCode, PLAY_SERVICES_RESOLUTION_REQUEST).show();
            } else {
                Toast.makeText(context, "Play Services not available", Toast.LENGTH_SHORT).show();
            }
            return false;
        }
        return true;
    }

    private synchronized void buildGoogleApiClient(Context context) {
        mLocationRequest = LocationRequest.create();
        mLocationRequest.setInterval(INTERVAL);
        mLocationRequest.setFastestInterval(FASTEST_INTERVAL);
        mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);

        mGoogleApiClient = new GoogleApiClient.Builder(context)
                .addApi(LocationServices.API)
                .addConnectionCallbacks(this)
                .addOnConnectionFailedListener(this)
                .build();
        mGoogleApiClient.connect();
    }

    @Override
    public void onLocationChanged(Location location) {
        Log.d(TAG, "onLocationChanged() returned: " + location);
        mCurrentLocation = location;
        mLastUpdateTime = DateFormat.getTimeInstance().format(new Date());
        //updateUI();
        latlongRec.onLocationReceived(location);

        stopLocationUpdates();   //-- stop if needs gps only once
        disconnectGoogleClient();
    }

    @Override
    public void onConnected(@Nullable Bundle bundle) {
        startLocationUpdates();
    }

    private void startLocationUpdates() throws SecurityException {
        if (mGoogleApiClient.isConnected()) {
            LocationServices.FusedLocationApi.requestLocationUpdates(mGoogleApiClient, mLocationRequest, this);
        }
    }

    public void disconnectGoogleClient() {
        if (mGoogleApiClient != null && mGoogleApiClient.isConnected()) {
            LocationServices.FusedLocationApi.removeLocationUpdates(mGoogleApiClient, this);
            mGoogleApiClient.disconnect();
        }
    }

    @Override
    public void onConnectionSuspended(int i) {
        Log.d(TAG, "onConnectionSuspended() returned: ");
    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
        Log.d(TAG, "onConnectionFailed() returned: ");
    }

    protected void stopLocationUpdates() {
        LocationServices.FusedLocationApi.removeLocationUpdates(mGoogleApiClient, this);
        mGoogleApiClient.disconnect();
        Log.e(TAG, "Location update stopped .......................");
    }

}

File: android_samples_kotlin/02 LocationDemo/app/src/main/java/com/msnegi/locationdemo/LocManager.kt
package com.msnegi.locationdemo

import android.content.Context
import android.location.Location
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.LocationListener
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import java.text.DateFormat
import java.util.Date

class LocManager : LocationListener, GoogleApiClient.ConnectionCallbacks,
    GoogleApiClient.OnConnectionFailedListener {
    var mGoogleApiClient: GoogleApiClient? = null
    private var mContext: Context? = null
    var mLocationRequest: LocationRequest? = null
    var mCurrentLocation: Location? = null
    var mLastUpdateTime: String = ""
    var latlongRec: LatLongReceived? = null

    fun connectGoogleClient(context: Context, latlongRec: LatLongReceived) {
        this.mContext = context
        this.latlongRec = latlongRec
        if (isGooglePlayServiceAvailable(context)) {
            buildGoogleApiClient(context)
        }
    }

    @Synchronized
    private fun buildGoogleApiClient(context: Context) {
        mLocationRequest = LocationRequest.create()
        mLocationRequest!!.setInterval(INTERVAL)
        mLocationRequest!!.setFastestInterval(FASTEST_INTERVAL)
        mLocationRequest!!.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)

        mGoogleApiClient = GoogleApiClient.Builder(context)
            .addApi(LocationServices.API)
            .addConnectionCallbacks(this)
            .addOnConnectionFailedListener(this)
            .build()
        mGoogleApiClient!!.connect()
    }

    override fun onLocationChanged(location: Location) {
        Log.d(TAG, "onLocationChanged() returned: " + location)
        mCurrentLocation = location
        mLastUpdateTime = DateFormat.getTimeInstance().format(Date())
        //updateUI();
        latlongRec!!.onLocationReceived(location)

        stopLocationUpdates() //-- stop if needs gps only once
        disconnectGoogleClient()
    }

    override fun onConnected(bundle: Bundle?) {
        startLocationUpdates()
    }

    @Throws(SecurityException::class)
    private fun startLocationUpdates() {
        if (mGoogleApiClient!!.isConnected()) {
            LocationServices.FusedLocationApi.requestLocationUpdates(
                mGoogleApiClient!!,
                mLocationRequest!!,
                this
            )
        }
    }

    fun disconnectGoogleClient() {
        if (mGoogleApiClient != null && mGoogleApiClient!!.isConnected()) {
            LocationServices.FusedLocationApi.removeLocationUpdates(mGoogleApiClient!!, this)
            mGoogleApiClient!!.disconnect()
        }
    }

    override fun onConnectionSuspended(i: Int) {
        Log.d(TAG, "onConnectionSuspended() returned: ")
    }

    override fun onConnectionFailed(connectionResult: ConnectionResult) {
        Log.d(TAG, "onConnectionFailed() returned: ")
    }

    protected fun stopLocationUpdates() {
        LocationServices.FusedLocationApi.removeLocationUpdates(mGoogleApiClient!!, this)
        mGoogleApiClient!!.disconnect()
        Log.e(TAG, "Location update stopped .......................")
    }

    companion object {
        private val TAG: String = LocManager::class.java.simpleName

        @Volatile
        private var instance: LocManager? = null

        fun getInstance(): LocManager {
            return instance ?: synchronized(this) {
                instance ?: LocManager().also { instance = it }
            }
        }


        const val PLAY_SERVICES_RESOLUTION_REQUEST: Int = 9000
        private val INTERVAL = (1000 * 10).toLong()
        private val FASTEST_INTERVAL = (1000 * 5 //benificial in case of services only
                ).toLong()

        fun isGooglePlayServiceAvailable(context: Context): Boolean {
            return checkPlayServices(context)
        }

        fun checkPlayServices(context: Context): Boolean {
            val gApi = GoogleApiAvailability.getInstance()
            val resultCode = gApi.isGooglePlayServicesAvailable(context)
            if (resultCode != ConnectionResult.SUCCESS) {
                if (gApi.isUserResolvableError(resultCode)) {
                    gApi.getErrorDialog(
                        context as android.app.Activity,
                        resultCode,
                        LocManager.Companion.PLAY_SERVICES_RESOLUTION_REQUEST
                    )!!.show()
                } else {
                    Toast.makeText(context, "Play Services not available", Toast.LENGTH_SHORT)
                        .show()
                }
                return false
            }
            return true
        }
    }
}
